### PR TITLE
Conseil 168 adding endpoints library for openAPI doc generation

### PR DESCRIPTION
**Pull request by piotrkosecki**
First of all, I want to mention that I had to make some workarounds because `endpoints` library used here was lacking some features or `circe` library really insisted on type safety which I needed to provide with various encoders/decoders. Some of them are left not implemented as they are not needed.

I've added to the resources files for swagger UI which is available at `http://localhost:1337/docs`, openAPI doc is available at `http://localhost:1337/openapi.json`. 

Not everything works as we would wish - as we are returning in many endpoints `Map[String, Any]` we cannot define type properly in openAPI with this lib, at least not yet.

I'll need also to add javadocs, and create better structure for all the traits/classes I've implemented, but I think it will be more obvious as I'll start implementing other endpoints.

### CODE REVIEW

File: src/main/scala/tech/cryptonomic/conseil/tezos/ApiOperations.scala [hunk 1]
```diff
@@ -144,7 +144,14 @@ object ApiOperations extends DataOperations {
             set = accountDelegates.toList
           )
         ).filter(_.set.nonEmpty),
-        limit = limit.getOrElse(DataTypes.defaultLimitValue)
+        limit = limit.getOrElse(DataTypes.defaultLimitValue),
+        orderBy = sortBy.map { o =>
+          val direction = order match {
+            case Some(value) if value == AscendingSort => OrderDirection.asc
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/ApiOperations.scala]: you are right, good catch :)

File: src/main/scala/tech/cryptonomic/conseil/tezos/ApiOperations.scala [hunk 2]
```diff
@@ -254,15 +261,15 @@ object ApiOperations extends DataOperations {
     * @param ec ExecutionContext needed to invoke the data fetching using async results
     * @return Operation group along with associated operations and accounts
     */
-  def fetchOperationGroup(operationGroupHash: String)(implicit ec: ExecutionContext): Future[Option[Map[Symbol, Any]]] = {
+  def fetchOperationGroup(operationGroupHash: String)(implicit ec: ExecutionContext): Future[Option[Map[String, Any]]] = {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/ApiOperations.scala]: Sure, introduced type `AnyMap`

File: src/main/scala/tech/cryptonomic/conseil/Conseil.scala [hunk 3]
```diff
@@ -51,16 +53,26 @@ object Conseil extends App with LazyLogging with EnableCORSDirectives with Conse
...
             complete("Supported methods : GET and POST.")
           }
         }
-      }
+      } ~ pathPrefix("docs") {
+        pathEndOrSingleSlash {
+          getFromResource("web/index.html")
+        }
+      } ~
+        pathPrefix("swagger-ui") {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/Conseil.scala]: sure, refactored, but intellij is not happy about that 😄

File: src/main/scala/tech/cryptonomic/conseil/routes/Data.scala [hunk 4]
```diff
@@ -26,102 +25,103 @@ object Data {
   * @param apiExecutionContext   is used to call the async operations exposed by the api service
   */
 class Data(config: PlatformsConfiguration, queryProtocolPlatform: DataPlatform)(implicit apiExecutionContext: ExecutionContext)
-  extends LazyLogging with RouteHandling with DatabaseApiFiltering with JacksonSupport {
+  extends LazyLogging with RouteHandling with DatabaseApiFiltering with DataHelpers {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/Data.scala]: I removed `RouteHandling` from here, but `avgFeesRoute` will be harder as there is no `average` operation implemented in query

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package tech.cryptonomic.conseil.routes.openapi
+
+trait TupleFlattenHelper {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala]: I used one which you provided, thanks!

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/EndpointsHelpers.scala [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package tech.cryptonomic.conseil.routes.openapi
+
+import endpoints.{InvariantFunctor, algebra}
+import tech.cryptonomic.conseil.tezos.ApiOperations.Filter
+
+trait EndpointsHelpers extends QueryStringLists with algebra.JsonSchemaEntities with Validation with TupleFlattenHelper {
+  implicit def qsInvFunctor: InvariantFunctor[QueryString]
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/EndpointsHelpers.scala]: went with what you provided, thanks!

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/DataHelpers.scala [hunk 7]
```diff
@@ -0,0 +1,89 @@
+package tech.cryptonomic.conseil.routes.openapi
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/DataHelpers.scala]: ok, moved

File: src/main/scala/tech/cryptonomic/conseil/Conseil.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/Conseil.scala]: Is it re-formatting your code?

We'll need to switch to scalafmt sooner or later

File: src/main/scala/tech/cryptonomic/conseil/routes/Data.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/routes/Data.scala]: I can imagine, but that query was actually less dependent on the common ApiFilter pattern, so I thought that we could make a custom query just for that, and then think of something better.
WDYT?

File: build.sbt [hunk 8]
```diff
@@ -2,14 +2,21 @@ name := "Conseil"
 scalaVersion := "2.12.8"
 
 val akkaHttpVersion = "10.1.0"
-val akkaVersion = "2.5.11"
+val akkaVersion = "2.5.19"
```
Comment (ivanopagano) [inline build.sbt]: no breaking change? semantic versioning seems to work for real...

File: src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    case x: java.lang.String => Json.fromString(x)
+    case x: java.lang.Integer => Json.fromInt(x)
+    case x: java.sql.Timestamp => Json.fromBigInt(x.getTime)
+    case x: java.lang.Boolean => Json.fromBoolean(x)
+    case x: scala.collection.immutable.Vector[Tables.OperationGroupsRow] => x.map(_.asJson(operationGroupsRowSchema.encoder)).asJson
+    case x: Tables.BlocksRow => x.asJson(blocksRowSchema.encoder)
+    case x => Json.fromString(x.toString)
+  }
+
+  override implicit def anySchema: JsonSchema[Any] = new JsonSchema[Any] {
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala]: Same considerations as above, a schema for `Any` type sounds like a runtime failure waiting to happen somewhere in the future.
Any alternative?

File: src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    override def encoder: Encoder[Any] = anyEncoder
+
+    override def decoder: Decoder[Any] =
+      (c: HCursor) => {
+        Right(c.value)
+      }
+  }
+
+  override implicit def queryResponseSchema: JsonSchema[List[Map[String, Option[Any]]]] =
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala]: I might suggest this type here: `Map[String, Option[Any]]` warrants his own type alias, and, in any case, I fail to see what are they representing?
I would love to see much more documentation here to figure out what and why is happening, I'm having a hard time trying to understand how things fit together.

File: src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    case x: java.lang.Boolean => Json.fromBoolean(x)
+    case x: scala.collection.immutable.Vector[Tables.OperationGroupsRow] => x.map(_.asJson(operationGroupsRowSchema.encoder)).asJson
+    case x: Tables.BlocksRow => x.asJson(blocksRowSchema.encoder)
+    case x => Json.fromString(x.toString)
+  }
+
+  override implicit def anySchema: JsonSchema[Any] = new JsonSchema[Any] {
+    override def encoder: Encoder[Any] = anyEncoder
+
+    override def decoder: Decoder[Any] =
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala]: let's try to figure this together tomorrow, this looks like pretending, this decoder is returning Json, and only if there's Json at the focus in the first place.

File: src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+      override def decoder: Decoder[List[Map[String, Option[Any]]]] = ???
+    }
+
+  override implicit def blocksByHashSchema: JsonSchema[AnyMap] = new JsonSchema[AnyMap] {
+    override def encoder: Encoder[AnyMap] = (a: AnyMap) => Json.obj(a.map {
+      case (k, v) => (k, v.asJson(anyEncoder))
+    }.toList: _*)
+
+    override def decoder: Decoder[AnyMap] = ???
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala]: Maybe we don't need decoders, and we can try to use JsonEntities without schema, let's talk about it too, I'm not sure how it's better to move here.

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/EndpointsHelpers.scala [hunk 13]
```diff
@@ -0,0 +1,56 @@
+package tech.cryptonomic.conseil.routes.openapi
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/routes/openapi/EndpointsHelpers.scala]: Will need to add docs

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/JsonSchemas.scala [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+  implicit def operationGroupsRowSchema: JsonSchema[Tables.OperationGroupsRow] =
+    genericJsonSchema[Tables.OperationGroupsRow]
+
+  implicit def avgFeeSchema: JsonSchema[AverageFees] =
+    genericJsonSchema[AverageFees]
+
+  implicit def anySchema: JsonSchema[Any]
+
+  implicit def queryResponseSchema: JsonSchema[List[Map[String, Option[Any]]]]
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/routes/openapi/JsonSchemas.scala]: Maybe `type QueryResponse = Map[String, Option[Any]]`?

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/OpenApiDoc.scala [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    List(
+      DocumentedParameter(name, false, docs, Schema.Array(implicitly[QueryStringParam[A]]))
+    )
+  )
+
+  override implicit def qsFunctor: Functor[QueryString] = new Functor[QueryString] {
+    override def map[From, To](f: OpenApiDoc.DocumentedQueryString)(map: From => To): OpenApiDoc.DocumentedQueryString = f
+  }
+
+  override implicit def timestampSchema: DocumentedJsonSchema = DocumentedJsonSchema.Primitive("integer")
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/routes/openapi/OpenApiDoc.scala]: Shouldn't be a long, at least?

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala]: I will find the original gist to refer to, so we know where everything came from.

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/Validation.scala [hunk 17]
```diff
@@ -0,0 +1,9 @@
+package tech.cryptonomic.conseil.routes.openapi
+
+import endpoints.algebra
+import endpoints.algebra.Documentation
+import tech.cryptonomic.conseil.generic.chain.DataTypes.QueryValidationError
+
+trait Validation extends algebra.Responses {
+  def validated[A](response: Response[A], invalidDocs: Documentation): Response[Either[List[QueryValidationError], A]]
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/routes/openapi/Validation.scala]: Maybe we should use cats Validated type here?

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/OpenApiDoc.scala (hunk 15 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/OpenApiDoc.scala]: https://swagger.io/docs/specification/data-models/data-types/
there is no type `long` in the openAPI spec

File: src/main/scala/tech/cryptonomic/conseil/Conseil.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/Conseil.scala]: I often use shortcuts for rearranging code and reformatting it. It reformats routes.

File: build.sbt (hunk 8 above)
Comment (piotrkosecki) [reply build.sbt]: from what I've tested it works just fine, I had some warnings about incompatibilities between akka libraries, that's why I introduced newer version.

File: src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+  override implicit def qsFunctor: Functor[QueryString] = new Functor[QueryString] {
+    override def map[From, To](f: QueryString[From])(map: From => To): QueryString[To] = new QueryString[To](
+      f.directive.map(map)
+    )
+  }
+
+  def anyEncoder: Encoder[Any] = (a: Any) => a match {
+    case x: java.lang.String => Json.fromString(x)
+    case x: java.lang.Integer => Json.fromInt(x)
+    case x: java.sql.Timestamp => Json.fromBigInt(x.getTime)
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala]: changed

File: build.sbt (hunk 8 above)
Comment (ivanopagano) [reply build.sbt]: Good, I actually did a through job of finding incompatibilities and will probably add something more to the updates later on!

File: src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait DataHelpers extends QueryStringListsServer with Validation with akkahttp.server.Endpoints
+  with akkahttp.server.JsonSchemaEntities with DataEndpoints {
+
+  import io.circe._
+  import io.circe.syntax._
+
+  override def validated[A](response: A => Route, invalidDocs: Documentation): Either[List[QueryValidationError], A] => Route = {
```
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala]: Let's say you can look at the doc about that and decide if it's useful or not

File: src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    case Right(None) => None
+  }
+
+  override implicit def qsFunctor: Functor[QueryString] = new Functor[QueryString] {
+    override def map[From, To](f: QueryString[From])(map: From => To): QueryString[To] = new QueryString[To](
+      f.directive.map(map)
+    )
+  }
+
+  def anyEncoder: Encoder[Any] = (a: Any) => a match {
```
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala]: Let's keep it like this as a starting point, and look for more room for improvement later on, if at all possible

File: src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala (hunk 9 above)
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala]: Let's record here that we agreed to stick with this, and see if we can do better later on

File: src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala (hunk 11 above)
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala]: This is more complicated, if we need to have decoders, at least we might try to figure out where and what is possible values to decode, so we can at least attempt to provide a reasonable solution

File: src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala (hunk 12 above)
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala]: Same as before, let's see if we can get rid of decoders, or at least understand what we expect to be passed in as `Any` from our client code

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala (hunk 16 above)
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala]: it's adapted from `travisbrown`'s solution.
My fork is here: https://gist.github.com/ivanopagano/282ec7112e29e63e603491961bdb2a5c

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/Validation.scala (hunk 17 above)
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/Validation.scala]: Same consideration as in previous comments

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala (hunk 16 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala]: added link to the `travisbrown`'s original solution in the comment

**Comment by piotrkosecki** [comment]
closing as this PR https://github.com/Cryptonomic/Conseil/pull/265 contains all the changes from here

**Pull request closed without merging**